lib/services/watcher.go

```diff
@@ -706,33 +731,35 @@ func (p *lockCollector) defineCollectorAsInitialized() {
 	})
 }
 
-// processEventAndUpdateCurrent is called when a watcher event is received.
-func (p *lockCollector) processEventAndUpdateCurrent(ctx context.Context, event types.Event) {
-	if event.Resource == nil || event.Resource.GetKind() != types.KindLock {
-		p.Log.Warningf("Unexpected event: %v.", event)
-		return
-	}
-
+// processEventsAndUpdateCurrent is called when a watcher event is received.
+func (p *lockCollector) processEventsAndUpdateCurrent(ctx context.Context, events []types.Event) {
 	p.currentRW.Lock()
 	defer p.currentRW.Unlock()
-	switch event.Type {
-	case types.OpDelete:
-		delete(p.current, event.Resource.GetName())
-		p.fanout.Emit(event)
-	case types.OpPut:
-		lock, ok := event.Resource.(types.Lock)
-		if !ok {
-			p.Log.Warningf("Unexpected resource type %T.", event.Resource)
-			return
+	for _, event := range events {
+		if event.Resource == nil || event.Resource.GetKind() != types.KindLock {
+			p.Log.Warningf("Unexpected event: %v.", event)
+			continue
 		}
-		if lock.IsInForce(p.Clock.Now()) {
-			p.current[lock.GetName()] = lock
+
+		switch event.Type {
+		case types.OpDelete:
+			delete(p.current, event.Resource.GetName())
 			p.fanout.Emit(event)
```

Comment:
Emit takes a slice of events, right? Would we gain performance by emitting them all at once after processing and filtering?

Reply:
For non-heartbeat resources I think if we get to a point where the difference matters something has likely gone horribly wrong, but it is probably best to use the more scalable pattern anyway so I've updated it to bulk-emit.